src/main/java/org/terasology/magicalstats/system/ManaAuthoritySystem.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.magicalstats.system;
+
+import gnu.trove.iterator.TFloatIterator;
+import gnu.trove.iterator.TIntIterator;
+import gnu.trove.list.TFloatList;
+import gnu.trove.list.TIntList;
+import org.junit.Before;
+import org.terasology.entitySystem.entity.EntityManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.ReceiveEvent;
+import org.terasology.entitySystem.prefab.Prefab;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterMode;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.entitySystem.systems.UpdateSubscriberSystem;
+import org.terasology.logic.characters.CharacterMovementComponent;
+import org.terasology.logic.characters.events.AttackEvent;
+import org.terasology.logic.characters.MovementMode;
+import org.terasology.logic.health.EngineDamageTypes;
+import org.terasology.logic.inventory.ItemComponent;
+import org.terasology.magicalstats.component.MagicalStatsComponent;
+import org.terasology.magicalstats.component.ManaComponent;
+import org.terasology.magicalstats.event.BeforeDrainedEvent;
+import org.terasology.magicalstats.event.BeforeManaRefillEvent;
+import org.terasology.magicalstats.event.BeforeManaRegenEvent;
+import org.terasology.magicalstats.event.DoDrainEvent;
+import org.terasology.magicalstats.event.DoManaRefillEvent;
+import org.terasology.magicalstats.event.DoManaRegenEvent;
+import org.terasology.magicalstats.event.FullManaEvent;
+import org.terasology.magicalstats.event.ManaChangedEvent;
+import org.terasology.magicalstats.event.OnDrainedEvent;
+import org.terasology.magicalstats.event.OnManaRefillEvent;
+import org.terasology.magicalstats.event.OnManaRegenEvent;
+import org.terasology.math.TeraMath;
+import org.terasology.protobuf.EntityData;
+import org.terasology.registry.In;
+import org.terasology.utilities.random.FastRandom;
+import org.terasology.utilities.random.Random;
+
+
+
+@RegisterSystem(RegisterMode.AUTHORITY)
+public class ManaAuthoritySystem extends BaseComponentSystem implements UpdateSubscriberSystem {
+    @In
+    private EntityManager entityManager;
+
+    @In
+    private org.terasology.engine.Time time;
+
+    @Override
+    public void update(float delta) {
```

Comment:
While this does work, and in some cases could maybe be justified from a gameplay perspective, *probably* mana regen doesn't need to be checked on every single game tick. This would be a prime candidate for optimization at some point in the future where likely mana regen could be checked far more rarely, maybe even less than once per second. Some systems do a throttle check as the first thing inside `update()` to only tick at some reduced rate, while more rare things can use the scheduler.